src/utils/cross_terms_combination.rs

```diff
@@ -168,6 +168,39 @@ fn merge<F: FieldUtils>(a: &[F], b: &[F], target: &mut [F], zip_with: &[F], bino
 
 }
 
+/// Struct to store last 2 layers of cross terms.
+struct Evals<F: FieldUtils> {
+    layouts_size: Vec<usize>,
+    data: [Vec<F>; 2],
+    current: usize
+}
+
+impl<F: FieldUtils> Evals<F> {
+    fn new(layouts: &Vec<Vec<EvalLayout>>, initial: Vec<F>) -> Self {
+        Evals {
+            data: [initial, vec![]],
+            layouts_size: layouts.iter().map(|l| l.total_size()).collect(),
+            current: 1
+        }
+    }
+
+    fn get_source_target_pair(&mut self, i: usize) -> (&mut [F], &mut [F]) {
+        let next = i + 1;
+        self.data[next % 2] = vec![F::ZERO; self.layouts_size[next]];
+        let ([a], [b]) = self.data.split_at_mut(1) else { unreachable!() };
+        self.current = next % 2;
+        return if next % 2 == 0 {
+            (b.as_mut_slice(), a.as_mut_slice())
+        } else {
+            (a.as_mut_slice(), b.as_mut_slice())
+        }
```

Comment:
I think what we actually want here is a custom Index/IndexMut implementation. It would be even more useful if we choose to go down the path of splitting `data` into two struct fields